docs/NotificaThor.1
.TH NotificaThor 1 "27 March 2013" "%%VERSION%%"


.SH NAME
NotificaThor - Highly configurable on screen displays for X.



.SH SYNOPSIS
notificathor
.BI "[-nvh] [-l " "logfile" "]"



.SH DESCRIPTION
NotificaThor is a lightweight and configurable program to display on-screen-notifications on events
like changing the brightness or ejecting a disk.
.br
It consists of a userspace daemon, that runs in the background and a command-line ulity
that sends messages to the daemon.



.SH OPTIONS
.TP
.B -h, --help
Display help message.

.TP
.B -V, --version
Display version info.

.TP
.B -v, --verbose
Print verbose messages.

.TP
.B -n, --nodaemon
Don't fork to the background and print to stderr instead of syslog.

.TP
.BI "-l " logfile ", --logfile=" logfile
Write messages to filename instead of syslog.



.SH CONFIGURATION FILE
The configuration file is placed in
.IR $XDG_CONFIG_HOME/NotificaThor/rc.conf .
NotificaThor monitors its config file and rereads it as soon as it is changed.
It follows a simple key=value scheme with the following keys:

.TP
.BI use_argb= bool
Determines wether to use an ARGB-visual for the OSD-window or not.
Using an ARGB-visual enables true transparency when using an compositing-manager
like compiz or xcompmgr.

.TP
.BI use_xshape= bool|'whole'
When 'whole' is specified, the whole popup will be click-through.
When 'true' is specified, only fully transparent areas will be click-through.

.TP
.BI default_theme= theme-name
Specifies a filename in the themes directory to use as the default theme.

.TP
.BI osd_default_x= [:]coordinate
Specifies the default x-coordinate relative to a centered window.
A preceeding colon specifies an absolute x-coordinate.

.TP
.BI osd_default_x= [:]coordinate
Specifies the default x-coordinate relative to a centered window.
A preceeding colon specifies an absolute x-coordinate.

.TP
.BI osd_default_timeout= seconds
Default timeout in seconds after which the window disapears.



.SH THEME FILES
Theme files reside in the
.IR $XDG_CONFIG_HOME/NotificaThor/themes/ "-directory."
However, one can put files into
.IR $XDG_CONFIG_HOME/NotificaThor/ ,
which contain additional settings, such as different images.
.br
These files can be called by the
.IR --popup "-option of"
.BR thor-cli (1).
.br
.RB "For further details on theming read " NotificaThor-themes (5).



.SH OTHER FILES
.TP
.I $XDG_CACHE_HOME/NotificaThor/PID
Contains process ID of currently running daemon.

.TP
.I $XDG_CACHE_HOME/NotificaThor/socket
Socket to communication between NotificaThor and thor-cli.



.SH SIGNALS
Sosd exits gracefully on reception of SIGTERM and SIGINT.


.SH BUGS
When no theme is specified, there will be an empty window.

.SH AUTHOR
Christian Weber

.SH VERSION
0.1.1

.SH SEE ALSO
thor-cli(1), NotificaThor-themes(5)


src/NotificaThor.c
/* ************************************************************* *\
 * NotificaThor.c                                                *
 *                                                               *
 * Project:     NotificaThor                                     *
 * Author:      Christian Weber (ChristianWeber802@gmx.net)      *
 *                                                               *
 * Description: Main function, event loop and event handling.    *
\* ************************************************************* */


#define _SOSD_MAIN_

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <syslog.h>
#include <sys/inotify.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/un.h>
#include <time.h>
#include <unistd.h>

#include "com.h"
#include "config.h"
#include "wins.h"
#include "NotificaThor.h"
#include "logging.h"


static sig_atomic_t sig_received = 0;
static char         *config_file;
static char         *pidfile;
static char         *socket_path;

int  xerror = 0;
char *config_path;
char *themes_path;


/*
 * Sets a timer to an amount of seconds
 * 
 * Parameters: timer   - The ID of the timer to set
 *             seconds - The time to set the timer to
 */
 static void
settimer( timer_t timer, double seconds)
{
	struct itimerspec t_spec =
	{
		.it_interval = { 0, 0 },
		.it_value    = { (time_t)seconds, (seconds - (time_t)seconds)*1000000000 }
	};
	
	timer_settime( timer, 0, &t_spec, NULL);
};


/*
 * Handles a message on the NotificaThor-socket
 * 
 * Parameters: sockfd - the filedescriptor of the socket
 *             timer  - the timer
 * 
 * Returns: 0 on success, -1 on error.
 */
static int
handle_message( int sockfd, timer_t timer)
{
	int            ret = 0;
	fd_set         set;
	int            clsockfd;
	thor_message   msg = {0};
	struct timeval timeout =
	{
		.tv_sec  = 1,
		.tv_usec = 0
	};
	
	
	/** Accepting connections **/
	if( (clsockfd = accept( sockfd, NULL, NULL)) == -1 ) {
		if( !sig_received )
			thor_errlog( LOG_CRIT, "Accepting connections from socket");
		return -1;
	}
	
	/** select to introduce 1 second connection timeout **/
	FD_ZERO( &set);
	FD_SET( clsockfd, &set);
	
	switch( select( clsockfd + 1, &set, NULL, NULL, &timeout) )
	{
		case 0:
			thor_log( LOG_ERR, "Connection timeout.");
			goto end;
		
		case -1:
			if( !sig_received )
				thor_errlog( LOG_CRIT, "select()");
			ret = -1;
			goto end;
	}
	
	/** reading **/
	if( receive_message( clsockfd, &msg) == -1) {
		thor_errlog( LOG_ERR, "Receiving message");
		goto end;
	}
	
	/** initializing the popup **/
	if( msg.timeout == 0 )
		msg.timeout = _osd_default_timeout;
		
	if( show_osd( &msg) == 0 )
		settimer( timer, msg.timeout);
		
	free_message( &msg);
	
  end:
	close( clsockfd);
	return ret;
};


/*
 * Initialize Inotify watch on config file.
 * 
 * Returns: Filedescriptor for Inotify or -1 on error.
 */
static int
init_inotify()
{
	int inofd = -1;
	
	
	if( (inofd = inotify_init()) != -1 ) {
		if( inotify_add_watch( inofd, config_file, IN_CLOSE_WRITE|IN_MODIFY) == -1 ) {
			thor_errlog( LOG_ERR, "Could not add watch on config file");
			close( inofd);
			inofd = -1;
		}
	}
	else
		thor_errlog( LOG_ERR, "Could not initialize inotify");
	
	return inofd;
};


/*
 * Handles Inotify-events on config file.
 * 
 * Parameters: fd       - Pointer to Inotify descriptor.
 *             ino_time - Pointer to the last time of Inotify-event.
 */
#define IN_EVENT_SIZE    (sizeof(struct inotify_event) + FILENAME_MAX + 1)
#define IN_BUFF_SIZE     (16*IN_EVENT_SIZE)
static void
handle_inotify( int *fd, time_t *ino_time)
{
	char buffer[IN_BUFF_SIZE];
	
	
	read( *fd, buffer, IN_BUFF_SIZE);
	/* check if the last inotify event was at least one second ago,
	 * because most text editors do strange things instead of simply
	 * writing text to a file...
	 */
	if( *ino_time < time( NULL) ) {
		thor_log( LOG_DEBUG, "Rereading config file...");
		sleep(1);
		parse_conf( config_file);
		query_extensions();
		*ino_time = time( NULL);
	}
	
	close( *fd);
	*fd = init_inotify();
}


/*
 * Checks for running instances of the daemon via PID file.
 * 
 * Returns: 0 if no instance is running, -1 if an instance
 *          is running or an error occurred.
 */
static int
check_for_running()
{
	pid_t pid;
	FILE  *fpid;
	char  string[10];
	char  *endptr;
	
	
	/** read existing file **/
	if( (fpid = fopen( pidfile, "r")) == NULL ) {
		if( errno == ENOENT )
			goto write_file;
		else {
			thor_errlog( LOG_CRIT, "Opening PID file");
			return -1;
		}
	}
	fgets( string, 9, fpid);
	fclose( fpid);
	if( (pid = strtol( string, &endptr, 10)) != 0 ) {
		if( kill( pid, 0) == 0 ) {
			thor_log( LOG_CRIT, "An instance of NotificaThor is already running ( PID = %d ).", pid);
			return -1;
		}
	}
	remove( socket_path);
	
  write_file:
	mkdir( pidfile, 0755);
	strcat( pidfile, "/"APP_NAME);
	mkdir( pidfile, 0700);
	strcat( pidfile, "/PID");
	
	if( (fpid = fopen( pidfile, "w")) == NULL ) {
		thor_errlog( LOG_CRIT, "Writing PID file");
		return -1;
	}
	fprintf( fpid, "%d", getpid());
	fclose( fpid);
	
	return 0;
};


/*
 * Signal handler
 */
static void
sig_handler( int sig)
{
	sig_received = sig;
	return;
};


/*
 * Handler for timeout.
 */
static void
timeout_handler()
{
	kill_osd();
	return;
};

		
/*
 * Create socket, timer, sighandler and play main loop.
 * 
 * Returns: 0 on success, -1 on error.
 */
static int
event_loop()
{
	int                 sockfd = 0;
	int                 inofd  = 0;
	struct sockaddr_un  saddr;
	struct sigaction    term_sa    = {{0}};
	struct sigevent     ev_timeout = {{0}};
	timer_t             timer;
	char                *env;
	time_t              ino_time  = time( NULL);
	
		
	/** generate config paths **/
	env = getenv( "XDG_CONFIG_HOME");
	if( !env || !*env ) {
		env = getenv( "HOME");
		
		config_path = (char*)malloc( strlen( env) + sizeof("/.config/"APP_NAME"/") + MAX_THEME_LEN);
		cpycat( cpycat( config_path, env), "/.config/"APP_NAME"/");
	}
	else {
		config_path = (char*)malloc( strlen( env) + sizeof("/"APP_NAME"/") + MAX_THEME_LEN);
		cpycat( cpycat( config_path, env), "/"APP_NAME"/");
	}
	config_file = (char*)malloc( strlen( config_path) + sizeof("rc.conf"));
	cpycat( cpycat( config_file, config_path), "rc.conf");
	
	themes_path = (char*)malloc( strlen( config_path) + sizeof("themes/") + MAX_THEME_LEN);
	cpycat( cpycat( themes_path, config_path), "themes/");
	
	/** generate cache paths **/
	env = getenv( "XDG_CACHE_HOME");
	socket_path = saddr.sun_path;
	if( !env || !*env ) {
		env = getenv( "HOME");
		
		cpycat( cpycat( socket_path, env), "/.cache");
	}
	pidfile = (char*)malloc( strlen( socket_path) + sizeof("/"APP_NAME"/PID"));
	cpycat( pidfile, socket_path);
	strcat( socket_path, "/"APP_NAME"/socket");
	
	/** install signalhandler **/
	term_sa.sa_handler = sig_handler;
	sigaction( SIGINT , &term_sa, NULL);
	sigaction( SIGTERM, &term_sa, NULL);
	sigaction( SIGHUP , &term_sa, NULL);
	
	/** check for running instances **/
	if( check_for_running() == -1 )
		return -1;
	
	/** parse config file **/
	parse_conf( config_file);
	
	/** install timer **/
	ev_timeout.sigev_notify = SIGEV_THREAD;
	ev_timeout.sigev_notify_function = timeout_handler;
	timer_create( CLOCK_REALTIME, &ev_timeout, &timer);
	
	/** prepare window **/
	if( prepare_x() == -1 )
		goto err;
	
	/** initialize inotify watch on config file **/
	inofd = init_inotify();
	
	/** opening socket **/
	if( (sockfd = socket( AF_UNIX, SOCK_STREAM, 0)) == -1 ) {
		thor_errlog( LOG_CRIT, "Creating socket");
		goto err_x;
	}
	
	saddr.sun_family = AF_UNIX;
	
	if( bind( sockfd, (struct sockaddr*)&saddr, sizeof(struct sockaddr_un)) == -1 ) {
		thor_errlog( LOG_CRIT, "Binding socket");
		goto err_x;
	}
	if( listen( sockfd, 5) == -1 ) {
		thor_errlog( LOG_CRIT, "Listening on socket");
		goto err_x;
	}
	
	/** event loop **/
	thor_log( LOG_DEBUG, "NotificaThor started (%d). Awaiting connections.", getpid());
	while( 1 )
	{
		fd_set set;
		
		
		FD_ZERO( &set);
		FD_SET( sockfd, &set);
		if( inofd != -1 )
			FD_SET( inofd, &set);
		
		if( select( sockfd + 1, &set, NULL, NULL, NULL) == -1 ) {
			if( errno != EINTR )
				thor_log( LOG_CRIT, "select()");
			goto err_x;
		}
		
		/** message via thor-cli **/
		if( FD_ISSET( sockfd, &set) ) {
			if( handle_message( sockfd, timer) == -1 )
				goto err_x;
		}
		
		/** changes in config file **/
		else if( FD_ISSET( inofd, &set) )
			handle_inotify( &inofd, &ino_time);
	}
		
	return 0;
	
	/** cleaning up **/
  err_x:
	cleanup_x();
  err:
	if( sig_received ) {
		if( xerror )
			thor_log( LOG_DEBUG, "X11: IO-Error ocurred (most likely X-Server exited).");
		else
			thor_log( LOG_DEBUG, "Received signal %s.", strsignal( sig_received));
	}
	thor_log( LOG_DEBUG, "Exiting NotificaThor...");
	close( sockfd);
	remove( socket_path);
	remove( pidfile);
	go_up( pidfile);
	remove( pidfile);
	free( config_path);
	free( config_file);
	free( themes_path);
	free( pidfile);
	timer_delete( timer);
	close_logger();
	
	return -1;
};


/*
 * MAIN FUNCTION.
 * Do option parsing, setup logging method and fork
 */
int
main( int argc, char *argv[])
{
	char opt;
	char *logfile       = NULL;
	int  logging_method = 0;
	
	const char          optstring[] = "l:vnhV";
	const struct option long_opts[] =
	{
		{ "logfile" , required_argument, NULL, 'l'},
		{ "verbose" , no_argument      , NULL, 'v'},
		{ "nodaemon", no_argument      , NULL, 'n'},
		{ "help"    , no_argument      , NULL, 'h'},
		{ "version" , no_argument      , NULL, 'V'},
		{ NULL      , 0                , NULL, 0  }
	};
	
	
	/** option parsing **/
	while( (opt = getopt_long( argc, argv, optstring, long_opts, NULL)) != -1 )
	{
		switch( opt )
		{
			case 'l':
				logfile = optarg;
				break;
			
			case 'v':
				logging_method |= LOGGER_DEBUG;
				break;
				
			case 'n':
				logging_method |= LOGGER_STDERR;
				break;
			
			case 'h':
				fputs( USAGE, stderr);
				return 0;
			
			case 'V':
				fputs( VERSION_STRING, stderr);
				return 0;
		}
	}
	
	setup_logger( logging_method, logfile);
	
	/** forking **/
	if( logging_method & LOGGER_STDERR ) {
		if( event_loop() == -1 )
			return 1;
	}
	else {
		switch( fork() )
		{
			case 0:			// child
				fclose( stdin);
				fclose( stdout);
				fclose( stderr);
				chdir( "/");
				
				if( event_loop() == -1 )
					return 1;
				break;
			
			case -1:		// forking failed
				thor_errlog( LOG_CRIT, "fork()");
				return 1;
		}
	}
	
	return 0;
};
